[Site Isolation] Fix broken frame tree traversals in BackForwardCache::trySuspendPage

charliewolfe opened the pull request:
#### 8ccab2d8ee4504192c25f92a4f95ed29af6b0e80
<pre>
[Site Isolation] Fix broken frame tree traversals in BackForwardCache::trySuspendPage
<a href="https://bugs.webkit.org/show_bug.cgi?id=287834">https://bugs.webkit.org/show_bug.cgi?id=287834</a>
<a href="https://rdar.apple.com/145015809">rdar://145015809</a>

Reviewed by BJ Burg and Alex Christensen.

Remove a local frame early return and fix recursive functions so that they visit local frames that have
remote parents. No change in behavior while bfcache or site isolation is disabled.

* Source/WebCore/history/BackForwardCache.cpp:
(WebCore::firePageHideEventRecursively):
(WebCore::BackForwardCache::trySuspendPage):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::stopForBackForwardCache):
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::stopForBackForwardCache):
* Source/WebCore/page/Frame.h:

Canonical link: <a href="https://commits.webkit.org/290549@main">https://commits.webkit.org/290549@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aca90b612cd733b3d89b4fff4b8bbbfb89f1372a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90342 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9872 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45268 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95347 "Failed to checkout and rebase branch from PR 40736") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41118 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92394 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10259 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18214 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/95347 "Failed to checkout and rebase branch from PR 40736") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27123 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93343 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7855 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81952 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/95347 "Failed to checkout and rebase branch from PR 40736") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7581 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36317 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40249 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77905 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37387 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97171 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17530 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12921 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78535 "layout-tests (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17787 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77782 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77755 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22206 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20816 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10790 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14218 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17540 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22867 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17281 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20733 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19065 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebCore/history/BackForwardCache.cpp:
```diff
@@ -433,33 +433,30 @@ static void setBackForwardCacheState(Page& page, Document::BackForwardCacheState
+        // that the parent document's ignore-opens-during-unload counter should be incremented while the
+        // pagehide event is being fired in its subframes:
+        // https://html.spec.whatwg.org/multipage/browsers.html#unload-a-document
+        unloadCountIncrementer.emplace(document.get());
```
Neat.

burg commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2109,10 +2109,8 @@ void FrameLoader::stopForBackForwardCache()
-            localChild->protectedLoader()->stopForBackForwardCache();
-    }
+    for (RefPtr child = m_frame->tree().firstChild(); child; child = child->tree().nextSibling())
+        child->stopForBackForwardCache();
```
So is it a goal to remove use of Local/RemoteFrame dynamicDowncasts in preference for Frame methods that self-dispatch? (essentially the visitor pattern)

charliewolfe replied:
I think it depends on what the function does. For recursive functions like this one, having a function on Frame that skips over each RemoteFrame makes sense, but there may be other scenarios where it isn’t necessary.

burg reviewed (approved):
If this change causes layout test progression with site isolation enabled, it would be good to mention that. Or even "covered by existing tests" or "not covered by existing tests".

charliewolfe committed:
[Site Isolation] Fix broken frame tree traversals in BackForwardCache::trySuspendPage
https://bugs.webkit.org/show_bug.cgi?id=287834
rdar://145015809

Reviewed by BJ Burg and Alex Christensen.

Remove a local frame early return and fix recursive functions so that they visit local frames that have
remote parents. No change in behavior while bfcache or site isolation is disabled.

* Source/WebCore/history/BackForwardCache.cpp:
(WebCore::firePageHideEventRecursively):
(WebCore::BackForwardCache::trySuspendPage):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::stopForBackForwardCache):
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::stopForBackForwardCache):
* Source/WebCore/page/Frame.h:

Canonical link: https://commits.webkit.org/290549@main
